src/utils/normalize.ts

```diff
@@ -0,0 +1,94 @@
+import { add0x, isValidHexAddress } from '@metamask/utils';
+import type { Hex } from '@metamask/utils';
+import BN from 'bn.js';
+
+type EIP712Domain = {
+  verifyingContract: string;
+};
+
+type SignTypedMessageDataV3V4 = {
+  types: Record<string, unknown>;
+  domain: EIP712Domain;
+  primaryType: string;
+  message: unknown;
+};
+
+/**
+ * Normalizes the messageData for the eth_signTypedData
+ *
+ * @param messageData - The messageData to normalize.
+ * @returns The normalized messageData.
+ */
+export function normalizeTypedMessage(messageData: string) {
+  let data;
+  try {
+    data = parseTypedMessage(
+      messageData,
+    ) as unknown as SignTypedMessageDataV3V4;
+  } catch (e) {
+    // Ignore normalization errors and pass the message as is
+    return messageData;
+  }
+
+  const { verifyingContract } = data.domain ?? {};
+
+  if (!verifyingContract) {
+    return messageData;
+  }
+
+  data.domain.verifyingContract = normalizeContractAddress(verifyingContract);
+
+  return JSON.stringify(data);
+}
+
+/**
+ * Parses the messageData to obtain the data object for EIP712 normalization
+ *
+ * @param data - The messageData to parse.
+ * @returns The data object for EIP712 normalization.
+ */
+function parseTypedMessage(data: string) {
+  if (typeof data !== 'string') {
```

Comment:
Do we have a test for this?

Reply:
Done in https://github.com/MetaMask/eth-json-rpc-middleware/pull/309/commits/af2a8c25a11e53caed2f62ce00730bc271f3e839